library/alloc/src/collections/binary_heap.rs

```diff
@@ -1220,6 +1220,30 @@ impl<T> BinaryHeap<T> {
     pub fn clear(&mut self) {
         self.drain();
     }
+
+    /// Creates an iterator which uses a closure to determine if an element should be removed from the underlying vec.
+    /// If the closure returns true, then the element is removed and yielded. If the closure returns false, the element will remain in the binary heap and will not be yielded by the iterator.
+    /// # Examples
+    /// Splitting the binary heap into evens and odds, reusing the original allocation:
+    ///
+    /// ```
+    /// #![feature(binary_heap_drain_filter)]
+    /// use std::collections::BinaryHeap;
+    /// let mut a = BinaryHeap::from(vec![1, 2, 3, 4, 5, 6, 8, 9, 11, 13, 14, 15]);
+    /// let mut evens = a.drain_filter(|x| *x % 2 == 0).collect::<Vec<_>>();
+    /// evens.sort();
+    /// let odds = a.into_sorted_vec();
+    ///
+    /// assert_eq!(evens, vec![2, 4, 6, 8, 14]);
+    /// assert_eq!(odds, vec![1, 3, 5, 9, 11, 13, 15]);
+    /// ```
+    #[unstable(feature = "binary_heap_drain_filter", reason = "recently added", issue = "42849")]
+    pub fn drain_filter<F>(&mut self, filter: F) -> DrainFilter<'_, T, F>
+    where
+        F: FnMut(&mut T) -> bool,
+    {
+        DrainFilter { inner: self.data.drain_filter(filter) }
```

Comment:
This reuses the drain_filter implementation of `Vec`, which is problematic

https://github.com/rust-lang/rust/issues/43244#issuecomment-641638196
https://github.com/rust-lang/rfcs/pull/3288

This would propagate that behavior to yet another collection. This doesn't necessarily block this PR but I think it would at least block stabilization and tests may need some updates once that question is settled.

Reply:
The vec implementation has had more time and effort, so I reused it since it would be better to concentrate efforts on one implementation. Starting another implementation which would do very similar behavior seems wasteful.